modules/qtwidgets/properties/collapsiblegroupboxwidgetqt.cpp

```diff
@@ -412,4 +384,196 @@ const std::vector<PropertyWidgetQt*>& CollapsibleGroupBoxWidgetQt::getPropertyWi
+    // buttonLayout->addWidget(moveUpBtn);
+    // buttonLayout->addWidget(moveDownBtn);
+    buttonLayout->addStretch(1);
+
+    buttonWidget->setLayout(buttonLayout);
+    buttonWidget->setVisible(prop->getVisible());
+}
+
+void CollapsibleGroupBoxWidgetQt::insertProperty(Property* prop, size_t index) {
+    setUpdatesEnabled(false);
+    util::OnScopeExit enableUpdates([&]() { setUpdatesEnabled(true); });
+
+    const size_t insertIndex = std::min(index + 1, properties_.size());
+    const bool insertAtEnd = (insertIndex == properties_.size());
+
+    auto insertPoint = properties_.begin() + insertIndex;
+    auto widgetInsertPoint = propertyWidgets_.begin() + insertIndex;
+
+    properties_.insert(insertPoint, prop);
+    PropertyObserver::addObservation(prop);
+
+    auto factory = InviwoApplication::getPtr()->getPropertyWidgetFactory();
+    if (auto propertyWidget = static_cast<PropertyWidgetQt*>(factory->create(prop).release())) {
+        propertyWidgets_.insert(widgetInsertPoint, propertyWidget);
+
+        insertPropertyWidget(propertyWidget, insertAtEnd);
+    } else {
+        LogWarn("Could not find a widget for property: " << prop->getClassIdentifier());
+
+        // insert empty element to keep property widget vector in sync with property vector
+        propertyWidgets_.insert(widgetInsertPoint, nullptr);
+    }
+}
+
+void CollapsibleGroupBoxWidgetQt::insertPropertyWidget(PropertyWidgetQt* propertyWidget,
+                                                       bool insertAtEnd) {
+    auto addPropertyWidget = [&](QGridLayout* layout, int row, PropertyWidgetQt* widget) {
+        if (auto collapsibleWidget = dynamic_cast<CollapsibleGroupBoxWidgetQt*>(widget)) {
+            collapsibleWidget->setNestedDepth(this->getNestedDepth() + 1);
+            // make the collapsible widget go all the way to the right border
+            layout->addWidget(widget, row, 0, 1, 2);
+
+        } else {  // not a collapsible widget
+            widget->setNestedDepth(this->getNestedDepth());
+            // property widget should only be added to the left column of the layout
+            layout->addWidget(widget, row, 0);
+        }
+
+        if (isChildRemovable()) {
+            addButtonLayout(layout, row, widget->getProperty());
+        }
+
+        widget->setParentPropertyWidget(this);
+        widget->initState();
+    };
+
+    if (insertAtEnd) {
+        // append property widget
+        addPropertyWidget(propertyWidgetGroupLayout_, propertyWidgetGroupLayout_->rowCount(),
+                          propertyWidget);
+    } else {
+        // recreate property widget grid layout from scratch since QGridLayout does not support
+        // inserting rows.
+        //
+        // Assigning another widget to an already occupied grid cell is possible, but the
+        // widgets are put on top of each other.
+        auto layoutWidget = createPropertyLayoutWidget();
+        auto layout = static_cast<QGridLayout*>(layoutWidget->layout());
```

Comment:
This does seems complicated, could one not just remove all widgets from the layout and then add them back in the right order?

Reply:
But why move them, just remove all of them then addd them back, you only ever need one layout? Or am I missing something?

Reply:
We do
        for (auto w : widgets) {
            keyframesLayout_->removeWidget(w.second);
        }
        for (auto w : widgets) {
            keyframesLayout_->addWidget(w.second);
        } 

In the animation stuff, and that seems to work?

Reply:
That needs testing. There might be side effects like widgets being temporarily unparented. In addition, the widgets here might have another widget on the same row (for removal). But the widget does not know about it.

Reply:
`addWidget()` for a QGridLayout (http://doc.qt.io/qt-5/qgridlayout.html#addWidget) requires at least a row and column argument. Since QGridLayout does not support insertion and/or updating the grid rows, we would have to append the widgets always after the last row. Thereby increasing the row count every single time by n when we update the layout.

Reply:
The only difference to the solution used right now would be that `util::OnScopeExit freeOldLayout` would not be required.